### ISSUE TITLE
Pipelines are not guaranteed to be created for existing accounts.

### ISSUE THREAD

**Issue by balltrev** [aws-ia/terraform-aws-control_tower_account_factory#75]
Even if the account name is the same (for an already enrolled account), pipelines are not guaranteed to be created for existing accounts. It sometimes works, sometimes not.  I have tried to ascertain why the behaviour is intermittent, but so far I haven't been able to determine why.

_Originally posted by @PeterBengtson in https://github.com/aws-ia/terraform-aws-control_tower_account_factory/issues/31#issuecomment-1038065531_

**Comment by balltrev** [comment]
I believe the per-account customization pipeline (assuming that's the pipeline you're talking about) is created at the end of `aft-account-provisioning-framework` StateMachine execution. Could you check your `aft-account-provisioning-framework` executions for failures?

**Comment by PeterBengtson** [comment]
Undoubtedly. The question is of course why the state machine fails.

<img width="1080" alt="Screenshot 2022-02-15 at 09 30 23" src="https://user-images.githubusercontent.com/3897/154022869-32d4c6bc-04ea-454f-a1c7-d10eb2ab6cfd.png">

As you can see, out of a total of 45 runs, 12 have failed. Some of these failures are due to resource contention, as I mentioned in a separate issue.

<img width="1062" alt="Screenshot 2022-02-15 at 09 32 00" src="https://user-images.githubusercontent.com/3897/154023399-aa65441a-5ec1-4cd7-b74f-e0a3f83361e6.png">
 
Looking at the very first failure, we see this:

<img width="1088" alt="Screenshot 2022-02-15 at 09 32 32" src="https://user-images.githubusercontent.com/3897/154023511-68445226-6ba7-485c-ae71-6cdb34c0045c.png">

Hunting for the logs:

<img width="1105" alt="Screenshot 2022-02-15 at 09 32 40" src="https://user-images.githubusercontent.com/3897/154023558-1f428dda-ad6e-440a-b519-67907fc326cd.png">

We see this, an AccessDenied for a role you have updated a few days ago, so this particular error shouldn't remain now:

<img width="1147" alt="Screenshot 2022-02-15 at 09 33 17" src="https://user-images.githubusercontent.com/3897/154023690-86eaa09b-71e6-46fc-a372-052d466f995e.png">

All the remaining step function failures are due to resource contention, i.e. API quota exceeded.

The point is this: retriggering a failed state machine execution in the GUI does not rectify the situation; it does not create any pipeline when re-executed.

**Comment by PeterBengtson** [comment]
So: is it possible to re-trigger AFT to create the missing pipelines?

**Comment by PeterBengtson** [comment]
To continue: in a system of 30 accounts, 12 accounts were missing pipelines. I tried commenting out a module to get AFT to forget about it internally, then adding the module again to import the account and create its missing pipeline, but even though there were no execution errors, the pipeline was not created.

Is there any internal state in the DynamoDB tables that needs to be cleared out manually?

**Comment by PeterBengtson** [comment]
Also, the same thing seems to happen with the CT-specific accounts such as Log Archive and Audit. These two when imported to AFT never get any pipelines.

**Comment by PeterBengtson** [comment]
I just noticed that Service Catalog will update the account name if it contains spaces. This is an account name as it appears on the SSO login page:

<img width="344" alt="Screenshot 2022-02-16 at 21 12 01" src="https://user-images.githubusercontent.com/3897/154350758-df562d32-7a02-4415-afd8-a5efcadb1d75.png">

 It appears as follows in Control Tower:

<img width="1176" alt="Screenshot 2022-02-16 at 21 12 55" src="https://user-images.githubusercontent.com/3897/154350826-e34f2741-1e1a-4a34-af09-14d896c3f361.png">

But in Service Catalog, this is the name:

<img width="1284" alt="Screenshot 2022-02-16 at 21 11 47" src="https://user-images.githubusercontent.com/3897/154350875-e99c9419-af7e-413c-ac49-681263a5ade1.png">

Note that Service Catalog also has changed capitalisation.

If you by any chance are using CLI commands in CodeBuild and elsewhere and are not quoting the name argument, this might cause problems. 

This is an example of an account without a pipeline. Also, updating it does not change any tags, and it's not possible to move the account to another OU using AFT.

**Comment by balltrev** [comment]
Yeah, we're actively working on a more robust implementation for the ingestion process.

As for updating the tags not actually updating tags, it seems to me there must be an error with the `aft-account-request-processor` Lambda. Can you verify the CloudWatch logs show no errors, and that the new request isn't in your `aft-account-request-dlq.fifo` dead letter queue?

**Comment by PeterBengtson** [comment]
Just checked. There are no errors in the CloudWatch log group, and the DLQ is empty.

**Comment by PeterBengtson** [comment]
This is the state machine execution state:

<img width="1128" alt="Screenshot 2022-02-16 at 21 45 12" src="https://user-images.githubusercontent.com/3897/154355728-c97f174c-64dc-43d8-81c1-d786b4fe4141.png"

As you can see, all steps were executed successfully.

**Comment by PeterBengtson** [comment]
If I submit a very simple change, like changing a tag for the account, I see that the `ct-aft-request` pipeline is triggered correctly and that the CodeBuild generates Terraform code for 1 change. However, the aft-account-provisioning-framework state machine is never triggered. Could this be because the names differ?

**Comment by PeterBengtson** [comment]
On the other hand, I just noticed this when updating:

```
  # module.aft.module.aft_account_request_framework.aws_lambda_event_source_mapping.aft_account_request_action_trigger has changed
  ~ resource "aws_lambda_event_source_mapping" "aft_account_request_action_trigger" {
        id                                 = "8b0354ea-ce23-4360-bf28-66a8e4526759"
      ~ last_modified                      = "2022-02-12T16:54:00Z" -> "2022-02-16T20:40:00Z"
      ~ last_processing_result             = "OK" -> "PROBLEM: Function call failed"
```
I'll check it out.

**Comment by PeterBengtson** [comment]
Lo and behold:

```


2022-02-16T21:58:34.240+01:00Copy[ERROR] ClientError: An error occurred (ValidationException) when calling the DescribeProvisionedProduct operation: 1 validation error detected: Value 'k8s PoC 1' at 'name' failed to satisfy constraint: Member must satisfy regular expression pattern: [a-zA-Z0-9][a-zA-Z0-9._-]*Traceback (most recent call last):  File "/var/task/aft_account_request_action_trigger.py", line 44, in lambda_handler    new_account = not provisioned_product_exists(event_record)  File "/opt/python/lib/python3.8/site-packages/aft_common/account_request_framework.py", line 37, in provisioned_product_exists    provisioned_product = Account(  File "/opt/python/lib/python3.8/site-packages/aft_common/account.py", line 35, in provisioned_product    client.describe_provisioned_product(Name=self.account_name)  File "/opt/python/lib/python3.8/site-packages/botocore/client.py", line 386, in _api_call    return self._make_api_call(operation_name, kwargs)  File "/opt/python/lib/python3.8/site-packages/botocore/client.py", line 705, in _make_api_call    raise error_class(parsed_response, operation_name) | [ERROR] ClientError: An error occurred (ValidationException) when calling the DescribeProvisionedProduct operation: 1 validation error detected: Value 'k8s PoC 1' at 'name' failed to satisfy constraint: Member must satisfy regular expression pattern: [a-zA-Z0-9][a-zA-Z0-9._-]* Traceback (most recent call last):   File "/var/task/aft_account_request_action_trigger.py", line 44, in lambda_handler     new_account = not provisioned_product_exists(event_record)   File "/opt/python/lib/python3.8/site-packages/aft_common/account_request_framework.py", line 37, in provisioned_product_exists     provisioned_product = Account(   File "/opt/python/lib/python3.8/site-packages/aft_common/account.py", line 35, in provisioned_product     client.describe_provisioned_product(Name=self.account_name)   File "/opt/python/lib/python3.8/site-packages/botocore/client.py", line 386, in _api_call     return self._make_api_call(operation_name, kwargs)   File "/opt/python/lib/python3.8/site-packages/botocore/client.py", line 705, in _make_api_call     raise error_class(parsed_response, operation_name)
-- | --
```

It does crash when the name supplied to Service Catalog contains spaces.

**Comment by mathewmoon** [comment]
@PeterBengtson I was bit by that with the first account I created. The fix was to actually delete the record and then create a new one with the exact same data except for the name. If you just modify the record in Dynamodb then you will get a different exception stating that the account name can't be changed. With the version of AFT module that I was using at the time this worked without any side effects. Take this at your own peril though.

**Comment by PeterBengtson** [comment]
Great to get this verified - thank you, @mathewmoon! Rather than modifying the internal state of an application I didn't write, I'll wait for an official fix, especially since the team is looking at the whole naming issue anyway. @balltrev, any ETA? ;)

**Comment by balltrev** [comment]
We are looking to hopefully get this into our next release!

**Comment by PeterBengtson** [comment]
@balltrev Trevor, it's easy to forget one thing when concentrating on bugs and issues – many, many thanks to you and the team for creating AFT. I work as an AWS consultant, mostly with enterprise customers in fintech and security, and the AFT finally makes it possible to recommend Control Tower to this segment of users, as well as to startups. It's really Control Tower's missing piece and something I've waited for since CT was released. With both AWS and Terraform behind it, supporting it, the future looks bright. Thank you, and well done!

**Comment by balltrev** [comment]
@PeterBengtson we were able to reliably reproduce the Access Denied issue you were running into. It should also be included in the next release with the account request processor changes. Look for a release soon!

**Comment by balltrev** [comment]
Release [1.3.3](https://github.com/aws-ia/terraform-aws-control_tower_account_factory/releases/tag/1.3.3) should resolve this, we discovered some eventual consistency issues around creating and then assuming the role during the customization step function. Can you make a change to the account request repo for the accounts that failed? they should recover after updating to 1.3.3

**Comment by PeterBengtson** [comment]
They didn't. Even commenting out the module, pushing, then enabling it again in a new push did not fix the issue.

**Comment by fjromerom** [comment]
Hi team, I'm still facing this issue and we are unable to import some existing accounts to AFT.

Here is the CloudWatch logs of the Lambda function aft-account-request-processor:

``` 
{"time_stamp": "2022-02-21 00:18:30,083","log_level": "ERROR","log_message": "CT Request is not valid" }NoneType: None

{"time_stamp": "2022-02-21 00:18:30,105","log_level": "ERROR","log_message": {"FILE": "aft_account_request_processor.py","METHOD": "lambda_handler","EXCEPTION": ""} }Traceback (most recent call last): 

File "/var/task/aft_account_request_processor.py", line 65, in lambda_handler assert ct_request_is_validAssertionError

[ERROR] AssertionErrorTraceback (most recent call last):  File "/var/task/aft_account_request_processor.py", line 65, in lambda_handler assert ct_request_is_valid

{"time_stamp": "2022-02-21 00:20:07,123","log_level": "ERROR","log_message": "CT Request is not valid" }NoneType: None

{"time_stamp": "2022-02-21 00:20:07,123","log_level": "ERROR","log_message": {"FILE": "aft_account_request_processor.py","METHOD": "lambda_handler","EXCEPTION": ""} }Traceback (most recent call last):

File "/var/task/aft_account_request_processor.py", line 65, in lambda_handler assert ct_request_is_validAssertionError

[ERROR] AssertionErrorTraceback (most recent call last):  File "/var/task/aft_account_request_processor.py", line 65, in lambda_handler assert ct_request_is_valid
```

Specifically this account was created by AWS Control Tower, later the name and email were renamed and then incorporated to AFT.

I'm not sure it may be related but we also updated the AWS Control Tower version to the latest version 2.8.

Let me know if you need any additional information.

**Comment by stumins** [comment]
Hey @fjromerom, I've created a dedicated issue to track the problem we're having - please use this issue for problems relating to pipelines not being created.

**Comment by PeterBengtson** [comment]
Correction to my previous posting: it does actually seem that everything works now - but it took a while for the pipelines to execute. Every account now has a pipeline, and the tags are correct. I can't tell when things straightened out themselves, as I only noticed that the pipelines were here, and the tags correct, after moving accounts around the OUs directly in Service Catalog, which of course triggered the AFT pipelines after each completed OU move. The same thing with accounts with spaces in their names, such as "k8s PoC 1" - as part of moving it to the Sandbox OU I changed its name in AFT to "K8s-Poc-1" which is what it is named in Service Catalog.

Anyway, things are looking up! Thanks.
